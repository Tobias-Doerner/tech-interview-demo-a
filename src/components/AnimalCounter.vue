<template>
  <div class="flex justify-center items-center">
    <v-btn :disabled="value < 1" secondary @click="decrease">Remove</v-btn>
    <span class="w-52 text-center">{{ label }}</span>
    <v-btn secondary @click="increase">Add</v-btn>
  </div>
</template>

<script>
import VBtn from '@/components/VBtn'

export default {
  components: {
    VBtn
  },
  props: {
    species: {
      type: String,
      default: () => ''
    },
    value: {
      type: Number,
      default: 0,
      validator: (value) => value >= 0
    }
  },
  computed: {
    label() {
      return `Current ${this.species} count: ${this.value}`
    }
  },
  methods: {
    increase() {
      this.$emit('input', this.value + 1)
    },
    decrease() {
      if (this.value > 0) this.$emit('input', this.value - 1)
    }
  }
}
</script>
