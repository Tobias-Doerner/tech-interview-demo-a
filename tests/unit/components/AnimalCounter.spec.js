import { mount, shallowMount } from '@vue/test-utils'
import AnimalCounter from '@/components/AnimalCounter'

describe('AnimalCounter', () => {
  describe('props', () => {
    test('custom validator function of value property should work correctly', () => {
      const wrapper = shallowMount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 0
        }
      })
      const validator = wrapper.vm.$options.props.value.validator
      expect(validator(-999)).toBe(false)
      expect(validator(-1)).toBe(false)
      expect(validator(0)).toBe(true)
      expect(validator(1)).toBe(true)
      expect(validator(999)).toBe(true)
    })

    test('test default values', () => {
      const wrapper = shallowMount(AnimalCounter)
      expect(wrapper.vm.species).toEqual('')
      expect(wrapper.vm.value).toEqual(0)
    })
  })

  describe('computed properties', () => {
    test('label should be set correctly', () => {
      const wrapper = shallowMount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 123
        }
      })
      expect(wrapper.vm.label).toBe('Current cow count: 123')
    })
  })

  describe('methods', () => {
    test('increase method should emit input event correctly', async () => {
      const wrapper = shallowMount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 123
        }
      })
      wrapper.vm.increase()

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted('input')).toBeTruthy()
      expect(wrapper.emitted('input').length).toBe(1)
      expect(wrapper.emitted('input')[0]).toEqual([124])
    })

    test('decrease method should emit input event correctly', async () => {
      const wrapper = shallowMount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 123
        }
      })
      wrapper.vm.decrease()

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted('input')).toBeTruthy()
      expect(wrapper.emitted('input').length).toBe(1)
      expect(wrapper.emitted('input')[0]).toEqual([122])
    })

    test('decrease method should emit input event correctly', async () => {
      const wrapper = shallowMount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 0
        }
      })
      wrapper.vm.decrease()

      await wrapper.vm.$nextTick()

      expect(wrapper.emitted().input).toBeFalsy()
    })
  })

  describe('snapshot test', () => {
    test('should render correctly', () => {
      const wrapper = mount(AnimalCounter, {
        propsData: {
          species: 'cow',
          value: 123
        }
      })
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
