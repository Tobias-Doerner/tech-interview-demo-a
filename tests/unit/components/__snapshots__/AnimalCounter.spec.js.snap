// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimalCounter snapshot test should render correctly 1`] = `
<div
  class="flex justify-center items-center"
>
  <button
    class="base flex items-center m-1 secondary"
  >
    
    Remove
   
    <!---->
  </button>
   
  <span
    class="w-52 text-center"
  >
    Current cow count: 123
  </span>
   
  <button
    class="base flex items-center secondary"
  >
    Add 
    <!---->
  </button>
</div>
`;
